File: app/src/main/java/bf/fasobizness/bafatech/helper/MyFirebaseMessagingService.java
package bf.fasobizness.bafatech.helper;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import java.util.Map;
import java.util.Objects;
import java.util.Random;

import bf.fasobizness.bafatech.R;
import bf.fasobizness.bafatech.activities.user.messaging.ActivityMessage;
import bf.fasobizness.bafatech.activities.user.messaging.DefaultMessagesActivity;
import bf.fasobizness.bafatech.interfaces.API;
import bf.fasobizness.bafatech.models.MyResponse;
import bf.fasobizness.bafatech.utils.MySharedManager;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MyFirebaseMessagingService extends FirebaseMessagingService {
    private static final String TAG = "MyFirebaseMessaging";
    private LocalBroadcastManager broadcaster;

    @Override
    public void onCreate() {
        super.onCreate();
        broadcaster = LocalBroadcastManager.getInstance(this);
    }

    public MyFirebaseMessagingService() {
    }

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);

        if (remoteMessage.getData().size() > 0) {
            Log.d(TAG, "Message data payload: " + remoteMessage.getData());
            String title = Objects.requireNonNull(remoteMessage.getNotification()).getTitle();
            String body = Objects.requireNonNull(remoteMessage.getNotification()).getBody();
            showNotification(title, body, remoteMessage.getData());
            Intent intent = new Intent("MyData");
            intent.putExtra("body", remoteMessage.getNotification().getBody());
            intent.putExtra("msg", remoteMessage.getData().get("msg"));
            broadcaster.sendBroadcast(intent);
        }

        if (remoteMessage.getNotification() != null) {
            Log.d(TAG, "Message Notification Body: " + remoteMessage.getNotification().getBody());
        }

    }

    @Override
    public void onNewToken(@NonNull String token) {
        // Log.d(TAG, "Refreshed token: " + token);
        sendRegistrationToServer(token);
    }

    private void sendRegistrationToServer(String token) {
        API api = RetrofitClient.getClient().create(API.class);
        MySharedManager sharedManager = new MySharedManager(this);
        String user = sharedManager.getUser();
        String auth = sharedManager.getToken();
        if (!user.isEmpty() && !auth.isEmpty()) {
            Call<MyResponse> call = api.updateFCM(token, "Bearer " + auth);
            call.enqueue(new Callback<MyResponse>() {
                @Override
                public void onResponse(@NonNull Call<MyResponse> call, @NonNull Response<MyResponse> response) {
                    Log.d(TAG, response.toString());
                }

                @Override
                public void onFailure(@NonNull Call<MyResponse> call, @NonNull Throwable t) {
                    Log.d(TAG, t.toString());
                }
            });
        }
    }

    private void showNotification(String title, String body, Map<String, String> data) {

        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        String NOTIFICATION_CHANNEL_ID = "bf.fasobizness.bafatech";
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, "bf.fasobizness.bafatech",
                    NotificationManager.IMPORTANCE_DEFAULT);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(notificationChannel);
            }
        }

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID);

        notificationBuilder
                .setAutoCancel(true)
                .setDefaults(Notification.DEFAULT_ALL)
                .setWhen(System.currentTimeMillis())
                .setSmallIcon(R.drawable.megaphone)
                .setColor(getResources().getColor(R.color.colorPrimary))
                .setContentTitle(title)
                .setContentText(body);

        if (data.get("discussion_id") != null) {
            Intent i = new Intent(this, DefaultMessagesActivity.class);
            // i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            i.putExtra("discussion_id", data.get("discussion_id"));
            i.putExtra("receiver_id", data.get("receiver_id"));
            PendingIntent contentIntent = PendingIntent.getActivity(this, 0, i, PendingIntent.FLAG_UPDATE_CURRENT);
            notificationBuilder.setContentIntent(contentIntent);
        }

        if (notificationManager != null) {
            notificationManager.notify(new Random().nextInt(), notificationBuilder.build());
        }
    }
}

File: app/src/main/java/bf/fasobizness/bafatech/helper/MyFirebaseMessagingService.kt
package bf.fasobizness.bafatech.helper

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import bf.fasobizness.bafatech.R
import bf.fasobizness.bafatech.activities.user.messaging.DefaultMessagesActivity
import bf.fasobizness.bafatech.interfaces.API
import bf.fasobizness.bafatech.models.MyResponse
import bf.fasobizness.bafatech.utils.MySharedManager
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class MyFirebaseMessagingService : FirebaseMessagingService() {
    private lateinit var broadcaster: LocalBroadcastManager

    override fun onCreate() {
        super.onCreate()
        broadcaster = LocalBroadcastManager.getInstance(this)
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        if (remoteMessage.data.isNotEmpty()) {
            Log.d(TAG, "Message data payload: " + remoteMessage.data)
            val title = remoteMessage.notification?.title
            val body = remoteMessage.notification?.body
            showNotification(title, body, remoteMessage.data)
            val intent = Intent("MyData")
            intent.putExtra("body", remoteMessage.notification!!.body)
            intent.putExtra("msg", remoteMessage.data["msg"])
            broadcaster.sendBroadcast(intent)
        }
        if (remoteMessage.notification != null) {
            Log.d(TAG, "Message Notification Body: " + remoteMessage.notification!!.body)
        }
    }

    override fun onNewToken(token: String) {
        Log.d(TAG, "c: $token")
        sendRegistrationToServer(token)
    }

    private fun sendRegistrationToServer(token: String) {
        val api = RetrofitClient.getClient().create(API::class.java)
        val sharedManager = MySharedManager(this)
        val user = sharedManager.user
        val auth = sharedManager.token
        if (user.isNotEmpty() && !auth.isEmpty()) {
            val call = api.updateFCM(token, "Bearer $auth")
            call.enqueue(object : Callback<MyResponse?> {
                override fun onResponse(call: Call<MyResponse?>, response: Response<MyResponse?>) {
                    Log.d(TAG, response.toString())
                }

                override fun onFailure(call: Call<MyResponse?>, t: Throwable) {
                    Log.d(TAG, t.toString())
                }
            })
        }
    }

    private fun showNotification(title: String?, body: String?, data: Map<String, String?>) {
        val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        val NOTIFICATION_CHANNEL_ID = "bf.fasobizness.bafatech"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notificationChannel = NotificationChannel(NOTIFICATION_CHANNEL_ID, "bf.fasobizness.bafatech",
                    NotificationManager.IMPORTANCE_DEFAULT)
            notificationManager.createNotificationChannel(notificationChannel)
        }
        val notificationBuilder = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
        notificationBuilder
                .setAutoCancel(true)
                .setDefaults(Notification.DEFAULT_ALL)
                .setWhen(System.currentTimeMillis())
                .setSmallIcon(R.drawable.megaphone)
                //.setColor(resources.getColor(R.color.colorPrimary))
                .setContentTitle(title)
                .setContentText(body)
        if (data["discussion_id"] != null) {
            val i = Intent(this, DefaultMessagesActivity::class.java)
            // i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            i.putExtra("discussion_id", data["discussion_id"])
            i.putExtra("receiver_id", data["receiver_id"])
            val contentIntent = PendingIntent.getActivity(this, 0, i, PendingIntent.FLAG_UPDATE_CURRENT)
            notificationBuilder.setContentIntent(contentIntent)
        }
        notificationManager.notify(Random().nextInt(), notificationBuilder.build())
    }

    companion object {
        private const val TAG = "MyFirebaseMessaging"
    }
}